conjure-go-client/httpclient/internal/random_scorer.go

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"math/rand"
+	"net/http"
+)
+
+type randomScorer struct {
+	uris      []string
+	nanoClock func() int64
+}
+
+func (n *randomScorer) GetURIsInOrderOfIncreasingScore() []string {
+	uris := make([]string, len(n.uris))
+	copy(uris, n.uris)
+	rand.New(rand.NewSource(n.nanoClock())).Shuffle(len(uris), func(i, j int) {
```

Comment:
You shouldn't need to re-construct the rand source on every call because the seed value, `n.nanoClock()`, will never change as time goes on since we use `time.Now()`.

Reply:
I was concerned about the docs for NewSource:
```
// NewSource returns a new pseudo-random Source seeded with the given value.
// Unlike the default Source used by top-level functions, this source is not
// safe for concurrent use by multiple goroutines.
```
Would that be an issue when calling using a `Rand` based on the `Source` for different requests in parallel?

Reply:
Yes that is an issue given multiple requests can happen concurrently sharing the same middleware. However, the current behavior is just to randomize the given URIs with the seed value and then use those throughout. So to uphold the current behavior, you could move all of this logic into the constructor and statically return the copy of the uris since they don't change across requests.

If instead you wanted the URIs to be in random order for every request across the same instance of the random_scorer middleware, then yes we need to be defensive or re-construct like you did.

Reply:
I misread the seed function as a function pointer (`time.Now().UnixNano`) instead of the called function `time.Now().UnixNano()` so you are indeed correct here. No need to update